### PR TITLE
dialects: (ml_program) add ml_program.global and ml_program.global_load_const

### PR DESCRIPTION
**Pull request by kayode-gif**
see https://mlir.llvm.org/docs/Dialects/MLProgramOps/#ml_programglobal-ml_programglobalop.

I'm sure a custom printer and parser is required in order to get this going, however i'm not sure how to start with printer and parsing this one.

### CODE REVIEW

**Comment by superlopuh** [comment]
Can you please add `ml_program.global_load_const` also

**Comment by kayode-gif** [comment]
i think the issue is with ```value = parse_attribute()``` perhaps, since that attribute expects a type?

**Comment by kayode-gif** [comment]
> i think the issue is with `value = parse_attribute()` perhaps, since that attribute expects a type?

edit: i wasn't passing the a type attribute needed too

File: xdsl/dialects/ml_program.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+        else:
+            sym_visibility = StringAttr("")
+        if parser.parse_optional_keyword("mutable"):
+            mutability = UnitAttr()
+        sym_name = parser.parse_attribute()
+        if parser.parse_optional_punctuation("("):
+            value = parser.parse_attribute()
+            parser.parse_punctuation(")")
+        parser.parse_punctuation(":")
+        parser.parse_type()
```
Comment (superlopuh) [inline xdsl/dialects/ml_program.py:81]:
```suggestion
        type = parser.parse_type()
```

File: xdsl/dialects/ml_program.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+        global_op = cls(
+            sym_name=sym_name,
+            type_attr=value.type,
```
Comment (superlopuh) [inline xdsl/dialects/ml_program.py:85]:
```suggestion
            type_attr=type,
```

File: xdsl/dialects/ml_program.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+            sym_visibility = StringAttr("public")
+        elif parser.parse_optional_keyword("nested"):
+            sym_visibility = StringAttr("nested")
+        elif parser.parse_optional_keyword("private"):
+            sym_visibility = StringAttr("private")
+        else:
+            sym_visibility = StringAttr("")
+        is_mutable = UnitAttr()
+        sym_name = parser.parse_symbol_name().data
+        if parser.parse_optional_punctuation("("):
```
Comment (superlopuh) [inline xdsl/dialects/ml_program.py:86]: I don't think this is optional

Comment (kayode-gif) [reply xdsl/dialects/ml_program.py:86]: why isnt it, we can have a case like this: ```ml_program.global mutable @foobar : tensor<?xi32>```

File: xdsl/dialects/ml_program.py [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+    def parse(cls, parser: Parser) -> Self:
+        attrs = parser.parse_optional_attr_dict()
+        if parser.parse_optional_keyword("public"):
+            sym_visibility = StringAttr("public")
+        elif parser.parse_optional_keyword("nested"):
+            sym_visibility = StringAttr("nested")
+        elif parser.parse_optional_keyword("private"):
+            sym_visibility = StringAttr("private")
+        else:
+            sym_visibility = StringAttr("")
```
Comment (superlopuh) [inline xdsl/dialects/ml_program.py:92]: This isn't a legal value for visibility. In MLIR, this op has the `Symbol` interface, which we represent with `SymbolOpInterface`. Could you please add that to this op's traits? We'll need it later to find the relevant `global` op when using `global_load_context`

File: xdsl/dialects/ml_program.py [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+    def parse(cls, parser: Parser) -> Self:
+        attrs = parser.parse_optional_attr_dict()
+        if parser.parse_optional_keyword("public"):
+            sym_visibility = StringAttr("public")
+        elif parser.parse_optional_keyword("nested"):
+            sym_visibility = StringAttr("nested")
+        elif parser.parse_optional_keyword("private"):
+            sym_visibility = StringAttr("private")
+        else:
+            raise VerifyException("Expected 'public', 'private', or 'nested'")
```
Comment (superlopuh) [inline xdsl/dialects/ml_program.py:95]: Here we should raise a ParseError instead

Comment (kayode-gif) [reply xdsl/dialects/ml_program.py]: is there a difference with this and ```parser.raise_error```

Comment (superlopuh) [reply xdsl/dialects/ml_program.py]: `raise_error` is preferred, you're right

**Pull request merged**